tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py

```diff
@@ -0,0 +1,256 @@
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import state_ops
+from tensorflow.python.ops import variables
+from tensorflow.python.training import optimizer
+from tensorflow.python.training import training_ops
+
+
+class _RefVariableAsynchronousProcessor(optimizer._RefVariableProcessor):
+  """Processor for Variable."""
+  def update_op_asynchronous(self, optimizer, g, index):
+    if isinstance(g, ops.Tensor):
+      return optimizer._apply_dense(g, self._v, index)
+    else:
+      assert isinstance(g, ops.IndexedSlices), ("Gradient ", g, " is neither a "
+                                                "tensor nor IndexedSlices.")
+      # pylint: disable=protected-access
+      return optimizer._apply_sparse_duplicate_indices(g, self._v, index)
+
+
+class _DenseResourceVariableAsynchronousProcessor(optimizer._DenseResourceVariableProcessor):
+  """Processor for dense ResourceVariables."""
+  def update_op_asynchronous(self, optimizer, g, index):
+    # pylint: disable=protected-access
+    if isinstance(g, ops.IndexedSlices):
+      return optimizer._resource_apply_sparse_duplicate_indices(
+        g.values, self._v, g.indices, index)
+    return optimizer._resource_apply_dense(g, self._v, index)
+
+
+def _get_processor(v):
+  """The processor of v."""
+  if v.op.type == "VarHandleOp":
+    return _DenseResourceVariableAsynchronousProcessor(v)
+  if isinstance(v, variables.Variable):
+    return _RefVariableAsynchronousProcessor(v)
+  raise NotImplementedError("Trying to optimize unsupported type ", v)
+
+
+class DelayCompensatedGradientDescentOptimizer(optimizer.Optimizer):
+  """Optimizer that implements gradient descent with delay compensation.
+
+  See [Zheng, Shuxin, et al., 2016](https://arxiv.org/abs/1609.08326)
+  ([pdf](https://arxiv.org/pdf/1609.08326.pdf)).
+  """
+
+  def __init__(self, learning_rate, variance_parameter, num_workers=1,
+               use_locking=False, name="DelayCompensatedGradientDescent"):
+    """Construct a new gradient descent optimizer with delay compensation.
+
+    Args:
+      learning_rate: A Tensor or a floating point value.  The learning
+        rate to use.
+      variance_parameter: A Tensor or a floating point value. The lambda
+        value to use.
+      num_workers: A value to indicate number of workers computing gradients
+        asynchronously.
+      use_locking: If True use locks for update operations.
+      name: Optional name prefix for the operations created when applying
+        gradients. Defaults to "DelayCompensatedGradientDescent".
+      """
+    if num_workers <= 0:
+      raise ValueError("num_workers must be positive: %s" % num_workers)
+    super(DelayCompensatedGradientDescentOptimizer, self).__init__(
+          use_locking, name)
+    self._learning_rate = learning_rate
+    self._lambda = variance_parameter
+    self._num_workers = num_workers
+
+  def minimize(self, loss, global_step=None, var_list=None,
```

Comment:
Why not use the `minimize` from the parent class? It seems this is copied exactly?